File: src/main/java/arc/arc/ai/GPTManager.java
package arc.arc.ai;

import arc.arc.ARC;
import arc.arc.configs.Config;
import arc.arc.configs.ConfigManager;
import arc.arc.hooks.HookRegistry;
import arc.arc.hooks.citizens.CitizensHook;
import arc.arc.util.CooldownManager;
import arc.arc.util.TextUtil;
import lombok.extern.slf4j.Slf4j;
import net.kyori.adventure.text.Component;
import org.bukkit.Location;
import org.bukkit.entity.Player;
import org.bukkit.event.player.AsyncPlayerChatEvent;
import org.bukkit.scheduler.BukkitRunnable;
import org.bukkit.scheduler.BukkitTask;

import java.util.*;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentSkipListSet;

@Slf4j
public class GPTManager {

    private static final Map<String, GPTEntity> entities = new ConcurrentHashMap<>();
    private static final Config config = ConfigManager.of(ARC.plugin.getDataPath(), "gpt.yml");
    private static final Map<UUID, List<Conversation>> conversations = new ConcurrentHashMap<>();
    private static final Set<UUID> awaitingResponse = new ConcurrentSkipListSet<>();
    private static BukkitTask cleanupTask;

    public static void init() {
        entities.clear();
        conversations.clear();
        cancel();
        cleanupTask = new BukkitRunnable() {
            @Override
            public void run() {
                long now = System.currentTimeMillis();
                conversations.forEach((uuid, convs) -> {
                    boolean removed = convs.removeIf(c -> now - c.lastMessageTime > c.lifeTime);
                    if (removed) log.info("Removed expired conversations for player {}", uuid);
                    if (convs.isEmpty()) conversations.remove(uuid);
                });
            }
        }.runTaskTimerAsynchronously(ARC.plugin, 0, 20 * 30L);
    }

    public static void cancel() {
        if (cleanupTask != null && !cleanupTask.isCancelled()) cleanupTask.cancel();
    }

    public static CompletableFuture<Optional<String>> getResponse(Player player, String message,
                                                                  String id, String archetype) {
        String playerName = player.getName();
        GPTEntity entity = entities.computeIfAbsent(id, key -> new GPTEntity(config, archetype, id));
        if (!entity.archetype.equals(archetype)) {
            log.warn("Entity {} has different archetype {} than expected {}", id, entity.archetype, archetype);
        }
        return entity.getResponse(player.getUniqueId(), playerName, message);
    }

    public static void processMessage(AsyncPlayerChatEvent chatEvent) {
        if (awaitingResponse.contains(chatEvent.getPlayer().getUniqueId())) {
            return;
        }
        Player player = chatEvent.getPlayer();
        String message = chatEvent.getMessage();
        processMessage(message, player, true);
    }

    public static CompletableFuture<Void> processMessage(String message, Player player, boolean appendCancel) {
        List<Conversation> conv = conversations.get(player.getUniqueId());
        if (conv == null) return CompletableFuture.completedFuture(null);
        if (conv.isEmpty()) return CompletableFuture.completedFuture(null);

        Location playerLocation = player.getLocation();
        log.info("Looking for conversation for player {} at location {}", player.getName(), playerLocation);
        Conversation conversation;
        long now = System.currentTimeMillis();
        conversation = conv.stream()
                .filter(c ->
                        ((c.location.getWorld().getName().equals(playerLocation.getWorld().getName()) &&
                                c.location.distance(playerLocation) < c.radius) || (c.radius < 0))
                                && now - c.lastMessageTime < c.lifeTime)
                .findFirst()
                .orElse(null);
        if (conversation == null) {
            log.info("Player {} is not in range of any conversation", player.getName());
            return CompletableFuture.completedFuture(null);
        }
        log.info("Player {} is in range of conversation with entity {}", player.getName(), conversation.gptId);
        if (message.startsWith("!")) message = message.substring(1);
        return getResponseAndSend(player, message, conversation, appendCancel);
    }

    private static CompletableFuture<Void> getResponseAndSend(Player player, String message, Conversation conversation, boolean appendCancel) {
        awaitingResponse.add(player.getUniqueId());
        return getResponse(player, message, conversation.gptId, conversation.getArchetype())
                .thenAccept(response -> {
                    conversation.lastMessageTime = System.currentTimeMillis();
                    if (response.isEmpty()) {
                        log.warn("Empty response from GPT for player {}", player.getName());
                        awaitingResponse.remove(player.getUniqueId());
                        return;
                    }
                    Component responseMessage = formatMessage(response.get(), conversation, appendCancel);
                    new BukkitRunnable() {
                        @Override
                        public void run() {
                            displayChatBubble(response.get(), conversation);
                            if (conversation.privateConversation) {
                                player.sendMessage(responseMessage);
                            } else {
                                conversation.getLocation()
                                        .getNearbyPlayers(conversation.getRadius())
                                        .forEach(p -> p.sendMessage(responseMessage));
                            }
                            awaitingResponse.remove(player.getUniqueId());
                        }
                    }.runTask(ARC.plugin);
                });
    }

    private static void displayChatBubble(String message, Conversation conversation) {
        if (HookRegistry.citizensHook == null || conversation.npcId == null) return;
        if (message.length() > config.integer("max-bubble-length", 50)) {
            log.info("Message is too long for bubble: {}", message);
            return;
        }
        String s = TextUtil.mmToLegacy(message);
        List<CitizensHook.HologramLine> list = new ArrayList<>();
        String[] split = s.split("\n");
        for (String string : split) {
            list.add(new CitizensHook.HologramLine(string, config.integer("bubble-duration-ticks", 20 * 20)));
        }
        HookRegistry.citizensHook.addChatBubble(conversation.npcId, list);
    }

    private static Component formatMessage(String message, Conversation conversation, boolean appendCancel) {
        String format = config.string("message-format", "<gray><gold>%gpt_name%<gray> » <white>%message%");
        if (appendCancel)
            format += config.string("cancel-appendix", "\n<red><hover:show_text:'Нажмите, чтобы закончить'><click:run_command:/arc ai stop %id%>[Нажмите, чтобы закончить разговор]</click></hover>");
        format = format.replace("%gpt_name%", conversation.getTalkerName());
        format = format.replace("%message%", message);
        format = format.replace("%id%", conversation.gptId);
        return TextUtil.mm(format);
    }

    public static void startConversation(Player player, String id, String archetype, String talkerName,
                                         Location location, double radius, long lifeTime, String initialMessage,
                                         String endMessage, Integer npcId, boolean privateConversation) {
        String playerName = player.getName();
        entities.computeIfAbsent(id, key -> new GPTEntity(config, archetype, id));
        List<Conversation> convs = conversations.computeIfAbsent(player.getUniqueId(), key -> new ArrayList<>());
        if (convs.stream().anyMatch(c -> c.gptId.equals(id))) {
            log.info("Player {} already has conversation with entity {}", playerName, id);
            return;
        }
        convs.add(Conversation.builder()
                .playerUuid(player.getUniqueId())
                .location(location)
                .archetype(archetype)
                .radius(radius)
                .gptId(id)
                .lastMessageTime(System.currentTimeMillis())
                .lifeTime(lifeTime)
                .talkerName(talkerName)
                .npcId(npcId)
                .endMessage(endMessage)
                .privateConversation(privateConversation)
                .build());
        log.info("Player {} started conversation with entity {}", playerName, id);
        if (initialMessage != null) {
            getResponseAndSend(player, initialMessage, convs.getLast(), false);
        }
    }

    public static void endConversation(Player player, String id) {
        List<Conversation> convs = conversations.get(player.getUniqueId());
        if (convs == null) {
            return;
        }
        Conversation conversation = null;
        for (Conversation c : convs) {
            if (c.gptId.equals(id)) {
                conversation = c;
                break;
            }
        }
        log.info("Player {} ended conversation with entity {}", player.getName(), id);
        if (conversation == null) {
            log.warn("Player {} tried to end conversation with entity {} but it was not found", player.getName(), id);
            return;
        }
        if (conversation.endMessage != null) {
            processMessage(conversation.endMessage, player, false)
                    .thenAccept(a -> player.sendMessage(config.componentDef("end-message", "<red>Вы закончили разговор")));
        } else {
            player.sendMessage(config.componentDef("end-message", "<red>Вы закончили разговор"));
        }
        convs.remove(conversation);
    }

    public static void endAllConversations(Player player) {
        List<Conversation> convs = conversations.remove(player.getUniqueId());
        player.sendMessage(config.componentDef("end-all-message", "<red>Вы закончили все разговоры"));
        if (convs == null) {
            return;
        }
        log.info("Player {} ended all conversations", player.getName());
    }


    public static List<Conversation> getConversations(Player commandSender) {
        return conversations.getOrDefault(commandSender.getUniqueId(), List.of());
    }
}

File: src/main/kotlin/ru/arc/ai/GPTManager.kt
package ru.arc.ai

import org.bukkit.Location
import org.bukkit.entity.Player
import ru.arc.ai.config.LlmModuleConfig
import ru.arc.ai.config.NpcChatConfig
import ru.arc.ai.llm.ModerationOutcome
import ru.arc.ai.llm.ModerationService
import ru.arc.ai.llm.OpenRouterLlmClient
import ru.arc.ai.npc.NpcChatRpcClient
import ru.arc.core.ScheduledTask
import ru.arc.core.Tasks
import ru.arc.core.repeatingAsync
import ru.arc.core.ticks
import ru.arc.hooks.HookRegistry
import ru.arc.hooks.citizens.CitizensHook
import ru.arc.util.Logging.error
import ru.arc.util.Logging.info
import ru.arc.util.Logging.warn
import ru.arc.util.TextUtil
import ru.arc.util.TextUtils
import java.util.UUID
import java.util.concurrent.CompletableFuture
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.CopyOnWriteArrayList
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import java.util.concurrent.atomic.AtomicInteger

object GPTManager {
    private const val LLM_THREADS = 2

    private val entities = ConcurrentHashMap<String, GPTEntity>()
    private val conversations = ConcurrentHashMap<UUID, CopyOnWriteArrayList<Conversation>>()
    private val responseGate = AwaitingResponseGate()
    private val lastRequestAt = ConcurrentHashMap<UUID, Long>()
    private val threadNumber = AtomicInteger()
    private var cleanupTask: ScheduledTask? = null
    private var requestExecutor: ExecutorService? = null
    private var running = false

    private lateinit var npcChatConfig: NpcChatConfig
    private lateinit var moderationService: ModerationService
    private lateinit var npcChatRpcClient: NpcChatRpcClient

    @JvmStatic
    fun init(
        llmConfig: LlmModuleConfig,
        npcChatConfig: NpcChatConfig,
        llmClient: OpenRouterLlmClient,
        npcChatRpcClient: NpcChatRpcClient,
    ) {
        shutdown()
        this.npcChatConfig = npcChatConfig
        this.npcChatRpcClient = npcChatRpcClient
        val executor =
            Executors.newFixedThreadPool(LLM_THREADS) { runnable ->
                Thread(runnable, "arc-gpt-${threadNumber.incrementAndGet()}").apply {
                    isDaemon = true
                }
            }
        requestExecutor = executor
        moderationService = ModerationService(llmClient, llmConfig, executor)

        try {
            cleanupTask =
                Tasks.scheduler.repeatingAsync(period = (20 * 30L).ticks, delay = 0.ticks) {
                    val now = System.currentTimeMillis()
                    conversations.forEach { (uuid, convs) ->
                        val removed = convs.removeIf { now - it.lastMessageTime > it.lifeTime }
                        if (removed) info("Removed expired conversations for player {}", uuid)
                        if (convs.isEmpty()) conversations.remove(uuid)
                    }
                }
            running = true
        } catch (error: Exception) {
            executor.shutdownNow()
            requestExecutor = null
            throw error
        }
    }

    @JvmStatic
    fun shutdown() {
        running = false
        cancel()
        entities.clear()
        conversations.clear()
        responseGate.clear()
        lastRequestAt.clear()
        requestExecutor?.shutdownNow()
        requestExecutor = null
    }

    @JvmStatic
    fun cancel() {
        cleanupTask?.cancel()
        cleanupTask = null
    }

    @JvmStatic
    fun getResponse(
        player: Player,
        message: String,
        id: String,
        archetype: String,
    ): CompletableFuture<String?> {
        if (!running) return CompletableFuture.completedFuture(null)
        val entity = entities.computeIfAbsent(id) { createEntity(archetype, id, true) }
        if (entity.archetype != archetype) {
            warn("Entity {} has different archetype {} than expected {}", id, entity.archetype, archetype)
        }
        return entity.getResponse(player.uniqueId, player.name, message)
    }

    @JvmStatic
    fun moderationResponse(message: String): CompletableFuture<ModerResponse?> {
        if (!running || !::moderationService.isInitialized) {
            warn("AI moderation not initialized — skipping message")
            return CompletableFuture.completedFuture(null)
        }
        return moderationService.moderate(message).thenApply { result ->
            result ?: return@thenApply null
            val response =
                when (result.outcome) {
                    ModerationOutcome.OK -> ModerationResponse.OK
                    ModerationOutcome.BAD -> ModerationResponse.BAD
                    ModerationOutcome.UNKNOWN -> return@thenApply null
                }
            ModerResponse(response, result.comment)
        }.exceptionally { failure ->
            error("Error getting moderation response", failure)
            null
        }
    }

    private fun createEntity(archetype: String, id: String, useHistory: Boolean): GPTEntity =
        GPTEntity(npcChatConfig, npcChatRpcClient, archetype, id, useHistory)

    @JvmStatic
    fun processMessage(
        message: String,
        player: Player,
        appendCancel: Boolean,
    ): CompletableFuture<Void> {
        val conv = conversations[player.uniqueId] ?: return CompletableFuture.completedFuture(null)
        if (conv.isEmpty()) return CompletableFuture.completedFuture(null)

        val playerLocation = player.location
        val now = System.currentTimeMillis()
        val conversation =
            conv.firstOrNull { c ->
                val loc = c.location ?: return@firstOrNull false
                ((loc.world?.name == playerLocation.world?.name && loc.distance(playerLocation) < c.radius) || c.radius < 0) &&
                    now - c.lastMessageTime < c.lifeTime
            } ?: return CompletableFuture.completedFuture(null)

        val actualMessage = if (message.startsWith("!")) message.substring(1) else message
        if (actualMessage.length > npcChatConfig.maxInputChars) {
            player.sendMessage(npcChatConfig.tooLongMessage)
            return CompletableFuture.completedFuture(null)
        }
        val previousRequest = lastRequestAt[player.uniqueId]
        if (previousRequest != null && now - previousRequest < npcChatConfig.cooldownMillis) {
            return CompletableFuture.completedFuture(null)
        }
        lastRequestAt[player.uniqueId] = now
        return getResponseAndSend(player, actualMessage, conversation, appendCancel)
    }

    private fun getResponseAndSend(
        player: Player,
        message: String,
        conversation: Conversation,
        appendCancel: Boolean,
    ): CompletableFuture<Void> {
        val gptId = conversation.gptId ?: return CompletableFuture.completedFuture(null)
        val responseFuture =
            responseGate.run(player.uniqueId) {
                getResponse(
                    player,
                    message,
                    gptId,
                    conversation.archetype ?: "default",
                )
            } ?: return CompletableFuture.completedFuture(null)

        return responseFuture.handle { response, failure ->
            conversation.lastMessageTime = System.currentTimeMillis()
            Tasks.scheduler.runSync(
                Runnable {
                    if (!player.isOnline) return@Runnable
                    if (failure != null || response == null) {
                        warn("NPC dialogue failed for player={} persona={}", player.name, conversation.archetype, failure)
                        player.sendMessage(npcChatConfig.fallbackMessage)
                        return@Runnable
                    }
                    val responseMessage = formatMessage(response, conversation, appendCancel)
                    displayChatBubble(response, conversation)
                    if (conversation.privateConversation) {
                        player.sendMessage(responseMessage)
                    } else {
                        conversation.location?.getNearbyPlayers(conversation.radius)?.forEach {
                            it.sendMessage(responseMessage)
                        }
                    }
                },
            )
            null
        }
    }

    private fun displayChatBubble(message: String, conversation: Conversation) {
        if (HookRegistry.citizensHook == null || conversation.npcId == null) return
        if (message.length > npcChatConfig.maxBubbleLength) return
        val list =
            message.split("\n").map {
                CitizensHook.HologramLine("&f$it", npcChatConfig.bubbleDurationTicks)
            }
        HookRegistry.citizensHook?.addChatBubble(conversation.npcId, list)
    }

    private fun formatMessage(
        message: String,
        conversation: Conversation,
        appendCancel: Boolean,
    ) =
        TextUtil.mm(
            buildString {
                append(npcChatConfig.messageFormat)
                if (appendCancel) {
                    append(npcChatConfig.cancelAppendix)
                }
            }.replace("%gpt_name%", TextUtils.escapeMM(conversation.talkerName ?: ""))
                .replace("%message%", TextUtils.escapeMM(message))
                .replace("%id%", TextUtils.escapeMM(conversation.gptId ?: "")),
        )

    @JvmStatic
    fun startConversation(
        player: Player,
        id: String,
        archetype: String,
        talkerName: String,
        location: Location,
        radius: Double,
        lifeTime: Long,
        initialMessage: String?,
        endMessage: String?,
        npcId: Int?,
        privateConversation: Boolean,
    ) {
        if (!running) {
            warn("Cannot start GPT conversation before AI initialization")
            return
        }
        entities.computeIfAbsent(id) { createEntity(archetype, id, true) }
        val convs = conversations.computeIfAbsent(player.uniqueId) { CopyOnWriteArrayList() }
        if (convs.any { it.gptId == id }) return

        val conv =
            Conversation(
                playerUuid = player.uniqueId,
                location = location,
                archetype = archetype,
                radius = radius,
                gptId = id,
                lastMessageTime = System.currentTimeMillis(),
                lifeTime = lifeTime,
                talkerName = talkerName,
                npcId = npcId,
                endMessage = endMessage,
                privateConversation = privateConversation,
            )
        convs.add(conv)
        if (initialMessage != null) {
            val responseMessage = formatMessage(initialMessage, conv, appendCancel = true)
            displayChatBubble(initialMessage, conv)
            player.sendMessage(responseMessage)
        }
    }

    @JvmStatic
    fun endConversation(player: Player, id: String) {
        val convs = conversations[player.uniqueId] ?: return
        val conversation = convs.find { it.gptId == id } ?: return
        if (conversation.endMessage != null) {
            val responseMessage = formatMessage(conversation.endMessage, conversation, appendCancel = false)
            displayChatBubble(conversation.endMessage, conversation)
            player.sendMessage(responseMessage)
        }
        convs.remove(conversation)
        player.sendMessage(npcChatConfig.endMessage)
    }

    @JvmStatic
    fun endAllConversations(player: Player) {
        conversations.remove(player.uniqueId)
        lastRequestAt.remove(player.uniqueId)
        player.sendMessage(npcChatConfig.endAllMessage)
    }

    @JvmStatic
    fun getConversations(player: Player): List<Conversation> = conversations[player.uniqueId]?.toList() ?: emptyList()

    internal fun isRunning(): Boolean = running

    internal fun hasActiveExecutor(): Boolean = requestExecutor?.isShutdown == false
}

internal class AwaitingResponseGate {
    private val active = ConcurrentHashMap<UUID, ActiveRequest>()

    fun <T> run(
        playerId: UUID,
        operation: () -> CompletableFuture<T>,
    ): CompletableFuture<T>? {
        val request = ActiveRequest()
        if (active.putIfAbsent(playerId, request) != null) return null
        val future =
            try {
                operation()
            } catch (error: Exception) {
                CompletableFuture.failedFuture(error)
            }
        request.attach(future)
        return future.whenComplete { _, _ -> active.remove(playerId, request) }
    }

    fun isAwaiting(playerId: UUID): Boolean = active.containsKey(playerId)

    fun clear() {
        val requests = active.values.toList()
        active.clear()
        requests.forEach(ActiveRequest::cancel)
    }

    private class ActiveRequest {
        @Volatile
        private var future: CompletableFuture<*>? = null

        @Volatile
        private var cancelled = false

        fun attach(future: CompletableFuture<*>) {
            this.future = future
            if (cancelled) future.cancel(true)
        }

        fun cancel() {
            cancelled = true
            future?.cancel(true)
        }
    }
}
